spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Initializer.scala

```diff
@@ -0,0 +1,125 @@
+
+/**
+ * Initialization method to initialize bias and weight.
+ * The init method will be called in Module.reset()
+ */
+
+trait Initializer {
+
+  type Shape = Array[Int]
+
+  def init[T](weight: Tensor[T], bias: Option[Tensor[T]], fmt: String = "output_first")
+          (implicit ev: TensorNumeric[T]): Unit
+
+  protected def getFans(shape: Shape, dataFormat: String = "output_first"): (Int, Int) = {
+    val dims = shape.length
+    val (first, second) = dims match {
+      case 2 => (shape(0), shape(1))
+      case 4 =>
+        val receptiveFieldSize = shape(2) * shape(3)
+        (shape(0) * receptiveFieldSize, shape(1) * receptiveFieldSize)
+      case 5 =>
+        val receptiveFieldSize = shape(0) * shape(3) * shape(4)
+        (shape(1) * receptiveFieldSize, shape(2) * receptiveFieldSize)
+      case _ =>
+        val sqrtElem = Math.sqrt(shape.product).toInt
+        (sqrtElem, sqrtElem)
+    }
+    val (fanIn, fanOut) = if (dataFormat == "input_first") {
+      (first, second)
+    } else if (dataFormat == "output_first") {
+      (second, first)
+    } else {
+      throw new IllegalArgumentException(s"Invalid inputFormat: $dataFormat")
+    }
+    (Math.max(1, fanIn), Math.max(1, fanOut))
+  }
+}
+
+case object RandomUniform extends Initializer {
+
+  def init[T](weight: Tensor[T], bias: Option[Tensor[T]], fmt: String = "output_first")
+          (implicit ev: TensorNumeric[T]): Unit = {
+    val shape = weight.size()
+    val (fanIn, _) = getFans(shape, fmt)
+    val stdv = 1.0 / math.sqrt(fanIn)
+    weight.apply1(_ => ev.fromType[Double](RNG.uniform(-stdv, stdv)))
+    bias.foreach(_.apply1(_ => ev.fromType[Double](RNG.uniform(-stdv, stdv))))
+  }
+
+}
+
+/**
+ * In short, it helps signals reach deep into the network.
+ *
+ * During the training process of deep nn:
+ *        1. If the weights in a network start are too small,
+ *           then the signal shrinks as it passes through
+ *           each layer until it’s too tiny to be useful.
+ *
+ *        2. If the weights in a network start too large,
+ *           then the signal grows as it passes through each
+ *           layer until it’s too massive to be useful.
+ *
+ * Xavier initialization makes sure the weights are ‘just right’,
+ * keeping the signal in a reasonable range of values through many layers.
+ *
+ * More details on the paper
+ *  [Understanding the difficulty of training deep feedforward neural networks]
+ *  (http://jmlr.org/proceedings/papers/v9/glorot10a/glorot10a.pdf)
+ */
+case object Xavier extends Initializer {
+  def init[T](weight: Tensor[T], bias: Option[Tensor[T]], fmt: String = "output_first")
+          (implicit ev: TensorNumeric[T]): Unit = {
+    val shape = weight.size()
+    val (fanIn, fanOut) = getFans(shape)
+    val stdv = math.sqrt(6.0 / (fanIn + fanOut))
+    weight.apply1(_ => ev.fromType[Double](RNG.uniform(-stdv, stdv)))
+    bias.foreach(_.fill(ev.fromType(0)))
+  }
+
+}
+case object BilinearFiller extends Initializer {
```

Comment:
Please add a javadoc to explain the init algoirthm. Including its basic idea and paper(if possible)